cluster-provision/images/vm-image-builder/create-containerdisk.sh

```diff
@@ -6,90 +6,137 @@ if [ "$#" -ne 1 ]; then
+    fi
+
+    if ! [ -e "${file_name}" ]; then
+        # Download base VM image
+        curl -L "${url}" -o "${file_name}"
+    fi
+
+    echo "${file_name}"
+}
+
 function customize_image() {
-  local source_image=$1
-  local os_variant=$2
-  local customized_image=$3
-  local cloud_config=$4
+    local source_image=$1
+    local os_variant=$2
+    local customized_image=$3
+    local cloud_config=$4
+
+    # Backup the VM image and pass copy of the original image
+    # in case customizing script fail.
+    vm_image_copy="$(dirname "${customized_image}")/copy-${source_image}"
+    cp "${source_image}" "${vm_image_copy}"
 
-  # Backup the VM image and pass copy of the original image
-  # in case customizing script fail.
-  vm_image_copy="$(dirname "${customized_image}")/copy-${source_image}"
-  cp "${source_image}" "${vm_image_copy}"
+    # TODO: convert this script and its dependencies to container
+    ${CUSTOMIZE_IMAGE_SCRIPT} "${vm_image_copy}" "${os_variant}" "${customized_image}" "${cloud_config}"
 
-  # TODO: convert this script and its dependencies to container
-  ${CUSTOMIZE_IMAGE_SCRIPT} "${vm_image_copy}" "${os_variant}" "${customized_image}" "${cloud_config}"
+    # Backup no longer needed.
+    rm -f "${vm_image_copy}"
+}
 
-  # Backup no longer needed.
-  rm -f "${vm_image_copy}"
+function build_container() {
+    local -r build_directory=$1
+    local -r new_vm_image_name=$2
+    local -r arch=$3
+    local -r image_name=$4
+    local -r tag=$5
+
+    if [ $arch = "aarch64" ]; then
+        docker_build="docker buildx build --platform "linux/arm64" . -t ${image_name}:${tag}"
+    else
+        docker_build="docker build . -t ${image_name}:${tag}"
+    fi
+    DOCKER_CLI_EXPERIMENTAL=enabled $docker_build -f - <<END
+FROM scratch
+ADD --chown=107:107 ${build_directory}/${new_vm_image_name} /disk/
+END
 }
 
 function cleanup() {
-  if [ $? -ne 0 ]; then
-    rm -rf "${build_directory}"
-  fi
+    if [ $? -ne 0 ]; then
+        rm -rf "${build_directory}"
+    fi
 
-  rm -f "copy-${VM_IMAGE}"
+    rm -f "copy-${customized_image}"
 }
 
+trap 'cleanup' EXIT SIGINT
+
 export IMAGE_NAME=$1
 
 if [ -f "${SCRIPT_PATH}/${IMAGE_NAME}/create-image.sh" ]; then
     export TAG="devel"
-    readonly VM_IMAGE="provisioned-image.qcow2"
+    readonly customized_image="customized-image.qcow2"
     readonly build_directory="${SCRIPT_PATH}/${IMAGE_NAME}/build"
-    
+
     trap 'cleanup' EXIT SIGINT
-    
     mkdir -p "${build_directory}"
 
     pushd "${SCRIPT_PATH}/${IMAGE_NAME}"
       cleanup
       echo "Creating the image"
-      ./create-image.sh "${build_directory}/${VM_IMAGE}" 
+      ./create-image.sh "${build_directory}/${customized_image}"
 
       echo "Creating the containerdisk ..."
       docker build . -t ${IMAGE_NAME}:${TAG} -f - <<END
 FROM scratch
-ADD --chown=107:107 build/${VM_IMAGE} /disk/
+ADD --chown=107:107 build/${customized_image} /disk/
 END
     popd
-
-else 
-    export TAG=devel
+else
```

Comment:
The actual code I'm referring to is missing from the patch context but, since we're skipping our multi-arch setup entirely when a `create-image.sh` script is present, I think it would make sense to error out in that case if attempting to build for a non-native architecture. Or add the necessary glue code.

Additionally, since apparently the `alpine-make-vm-image` project lacks support for non-x86 architectures, the `alpine-cloud-init/create-image.sh` script should be updated to error out when attempting a build on any other architecture.

Note that there's a bunch of code in there that looks like it's intended to support foreign architecture builds, but I don't think it works? If it does, no reason to avoid multi-arch builds in this case. If it doesn't, it should be dropped to avoid confusion.

Reply:
I have add an error out in `alpine-cloud-init/create-image.sh` when it trying to build any no-x86 architecture image.